File: src/main/java/ru/darkkeks/vkmirror/tdlib/TelegramCredentials.java
package ru.darkkeks.vkmirror.tdlib;

public class TelegramCredentials {

    private int apiId;
    private String apiHash;

    private boolean isBot;
    private String phoneNumber;
    private String botToken;

    private TelegramCredentials(int apiId, String apiHash, String phoneNumber, String botToken, boolean isBot) {
        this.apiId = apiId;
        this.apiHash = apiHash;
        this.phoneNumber = phoneNumber;
        this.botToken = botToken;
        this.isBot = isBot;
    }

    public static TelegramCredentials phone(int apiId, String apiHash, String phoneNumber) {
        return new TelegramCredentials(apiId, apiHash, phoneNumber, null, false);
    }

    public static TelegramCredentials bot(int apiId, String apiHash, String botToken) {
        return new TelegramCredentials(apiId, apiHash, null, botToken, true);
    }

    public TdApi.Function getCredentialsFunction() {
        if (isBot) {
            return new TdApi.CheckAuthenticationBotToken(botToken);
        } else {
            return new TdApi.SetAuthenticationPhoneNumber(phoneNumber, false, false);
        }
    }

    public String getDataDirectory() {
        return "tdlib" + (isBot ? "-" + botToken.substring(0, botToken.indexOf(":")) : phoneNumber);
    }

    public int getApiId() {
        return apiId;
    }

    public String getApiHash() {
        return apiHash;
    }
}

File: src/main/java/ru/darkkeks/vkmirror/tdlib/TelegramCredentials.kt
package ru.darkkeks.vkmirror.tdlib

import ru.darkkeks.vkmirror.tdlib.internal.TdApi
import ru.darkkeks.vkmirror.tdlib.internal.TdApi.CheckAuthenticationBotToken
import ru.darkkeks.vkmirror.tdlib.internal.TdApi.SetAuthenticationPhoneNumber

fun botTelegramCredentials(apiId: Int, apiHash: String, botToken: String) =
        TelegramCredentials(apiId, apiHash, true, botToken = botToken)

fun userTelegramCredentials(apiId: Int, apiHash: String, phoneNumber: String) =
        TelegramCredentials(apiId, apiHash, false, phoneNumber = phoneNumber)

class TelegramCredentials(
        val apiId: Int,
        val apiHash: String,
        private val isBot: Boolean,
        private val phoneNumber: String? = null,
        private val botToken: String? = null) {

    val credentialsFunction: TdApi.Function get() = when {
        isBot -> CheckAuthenticationBotToken(botToken)
        else -> SetAuthenticationPhoneNumber(phoneNumber, TdApi.PhoneNumberAuthenticationSettings(false, true, false))
    }

    val dataDirectory: String get() = when {
        isBot -> "tdlib-$strippedToken"
        else -> "tdlib$phoneNumber"
    }

    private val strippedToken: String
        get() = botToken?.substring(0, botToken.indexOf(":")) ?: ""
}